### ISSUE TITLE
user defined fields

### ISSUE THREAD

**Issue by jakobroehrl** [nextcloud/deck#1833]
option to add special fields
 - text
 - longtext
 - dropdown
 - multiselect
 - checkbox, 
 - Mail
 - phone
 - URL

**Comment by goldengel** [comment]
I am looking for a "special field" feature too to have my DECK with a blank field with no mechanism. Just store the value I fill with. Is that possible? Is that under construction already with this issue?

**Comment by flogee** [comment]
i also hope this will be implemented!

**Comment by stefan-niedermann** [comment]
Please use **[GitHub Reactions](https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/)** 👍 if you are also interested in an issue to keep the history clean and readable. Comments which don't provide new content or information make it hard to understand and develop a new feature 🙂

**Comment by alexisig** [comment]
What would you think of the following feature breakdown? :slightly_smiling_face: 
- Possibility to set a list of custom fields globally in the app settings (so that custom fields don't need to be recreated on every board)
- Possibility to set a list of custom fields on a specific board (that overrides global fields if field names are the same)
- Custom card fields are displayed under the `Details` tab of a card
- Custom card fields are available via the API under a `customAttributes` object:
```
{
            "title": "Some card",
            "description": "",
            "stackId": 2,
            "type": "text",
            "lastModified": 1623776315,
            "lastEditor": null,
            "createdAt": 1613425886,
            "labels": null,
            "assignedUsers": [],
            "attachments": null,
            "attachmentCount": 1,
            "owner": "username,
            "order": 1,
            "archived": false,
            "duedate": "2021-06-06T04:00:00+00:00",
            "deletedAt": 0,
            "commentsUnread": 0,
            "commentsCount": 0,
            "id": 1,
            "ETag": "fba710d57eb7259fb1b995b1fdb3b9ed",
            "overdue": 3,
            "customAttributes": {
                "attribute":"value",
                "other_attribute": "value"
            }
        }
```

**Comment by stefan-niedermann** [comment]
Looks good to me :slightly_smiling_face:  - though the proposed API looks a bit rough at the moment. This will only allow key / value pairs and therefore not allow checkboxes, radio buttons and selects.

1. As a first step this is good, of course, but we should at least consider other types to be added in the future. So your attribute might look more something like this:
  ```json
"customAttributes": [
    {
        "id": "1",
        "label": "Start date",
        "type": "date",
        "value": "2020-02-02T19:00:00"
    }
]
```

2. Also we need to define how changes in the configuration behaves: What if a field gets deleted? What if it gets added or something changes about it? How does the API behave if a client posts a field that does no longer exist? Ignore it or throw an error?
3. What about i18n? Should the label of a new field be translatable? Maybe the [`Accept-Language`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Language) or [`Content-Language`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Language) header should be supported to no ship always all available translations?

**Comment by juliushaertl** [comment]
> As a first step this is good, of course, but we should at least consider other types to be added in the future. So your attribute might look more something like this

Yes, I think we should have types from the beginning, even if we start off with only a limited set or just text.

Apart from that it might be worth to think about how we can store types properly in the database so ideally we can also use proper sql queries if we have a date or integer we want to query for.

> What about i18n? Should the label of a new field be translatable? Maybe the Accept-Language or Content-Language header should be supported to no ship always all available translations?

I guess for user defined fields just having the label as entered is fine.

What I had in mind is that there could be also a list of predefined custom fields that would offer additional features, e.g. could be mapped to CalDAV. The start date would be a good example for that. For those translations would of course make sense, but for user defined storing custom translations seems a bit to much to me.

**Comment by mstoss** [comment]
> 
> 
> Looks good to me 🙂 - though the proposed API looks a bit rough at the moment. This will only allow key / value pairs and therefore not allow checkboxes, radio buttons and selects.
> 
>     1. As a first step this is good, of course, but we should at least consider other types to be added in the future. So your attribute might look more something like this:
> 
> 
> ```json
> "customAttributes": [
>   {
>       "id": "1",
>       "label": "Start date",
>       "type": "date",
>       "value": "2020-02-02T19:00:00"
>   }

Hello,
Is it really a good idea to have "Start date" as a customed value ? I am not sure because if "Start date" is completely integrated in "Deck" then it could be managed in the "Calendar App" as the "due date".

What do you think ?

Thank you !

**Comment by stefan-niedermann** [comment]
@mstoss this was just a random sample i chose as an example about how the data structure might look

**Comment by mstoss** [comment]
> 
> 
> @mstoss this was just a random sample i chose as an example about how the data structure might look

Thank you Stefan, 

Yes, I just commented here because in the issue #3146, there was a reference to this issue.

Maybe i chose the wrong issue to write my comment ...

**Comment by GuyInCorner** [comment]
I'm wondering, would it be possible to expose a unique Id of the card as part of this?  Would be handy to have the option.

**Comment by stevecopecs** [comment]
Hi people working on #1833
(I understand that I'm being an idiot, but I have tried to make sense of what's here, .. unsuccessfully)

I've been eagerly watching this as custom fields on cards appears to be a valuable enhancement.
I see various subsequent issues as having been opened / linked / closed, but can't get a coherent sense of how far the original request has progressed, ... or if there is anything I (as a non-coder user) can test and give feedback on. I REALLY don't want to frustrate anyone, would appreciate guidance as to how to get insight. Many thanks

**Comment by stevecopecs** [comment]
> Hi @stevecopecs,
> 
> The closed issues mentioned above are mostly duplicates closed in favour of this issue.
> 
> As for the state of this issue; it looks like the maintainers are interested, and there has been some unfinished discussion on how to implement this. However, Unless a developer gets involved to code this, it could be some time before it gets done.

Is there any way indication of cost / work estimation for a dev ?

**Comment by daudtivan** [comment]
Just to contribute with a mockup inspired on github cards style
![card-scale](https://user-images.githubusercontent.com/91741832/160852657-b9a11340-7e66-4fd9-b7b8-8f125288bebd.png)
Aligned all items in sequence at the bottom line
- tags
- likes/upvotes (attribute 1)
- effort estimate (attribute 2)

And I highly recommend a true designer to prepare a decent icon set (not mine ;-).
Hope it helps

**Comment by marcjoancr** [comment]
Any news on that request? 3 years since the creation and no updates.
